activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -608,10 +608,10 @@ def primary_keys(table_name) # :nodoc:
       end
 
       def case_sensitive_comparison(table, attribute, column, value)
-        if value.nil? || column.case_sensitive?
-          super
-        else
+        if !value.nil? && column.collation && !column.case_sensitive?
```

Comment:
If `case_sensitive?` is modified to `!collation || ..`, it returns always `true` in the case of **C**.

But it will break `can_perform_case_insensitive_comparison_for?`.
https://github.com/rails/rails/blob/c57e7239a8b82957bcb07534cb7c1a3dcef71864/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb#L619

`can_perform_case_insensitive_comparison_for?` should return `true` in the case of only **B**.